### PR TITLE
Add /health endpoint

### PR DESCRIPTION
**Pull request by S7evinK**
### Pull Request Checklist

Partly solves #1429 (checking the http connection to other components not yet included)
Pings the database where applicable. 

* [ ] I have added any new tests that need to pass to `sytest-whitelist` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/docs/CONTRIBUTING.md#sign-off)

### CODE REVIEW

**Review by Kegsay — CHANGES_REQUESTED** [review]
It'd probably be nicer to fold this into `base.SetupAndServeHTTP` as we always call that immediately after adding the health check. Also, it feels slightly wrong to have another copy of the `sql.DB` open just for calling `Ping()` on it. I'm unsure of the performance implications of doing this.

**Comment by S7evinK** [comment]
Makes sense to use the same `sql.DB`
I've changed the `sqlutil.Open` function to return an existing connection. Not sure if we need different connections in other cases.

File: internal/sqlutil/trace.go [hunk 1]
```diff
@@ -96,10 +96,18 @@ func trackGoID(query string) {
...
 
+var dbConns = make(map[string]*sql.DB)
+
 // Open opens a database specified by its database driver name and a driver-specific data source name,
 // usually consisting of at least a database name and connection information. Includes tracing driver
 // if DENDRITE_TRACE_SQL=1
 func Open(dbProperties *config.DatabaseOptions) (*sql.DB, error) {
+	// reuse connection only if not a sqlite3 test database
+	isTestDB := strings.HasSuffix(string(dbProperties.ConnectionString), "_test.db")
+	if conn, ok := dbConns[string(dbProperties.ConnectionString)]; !isTestDB && ok {
```
Comment (Kegsay) [inline internal/sqlutil/trace.go:107]: We've managed this long without leaking test-only code into the main codebase. Why is this necessary?

File: cmd/dendrite-room-server/main.go [hunk 2]
```diff
@@ -36,5 +36,6 @@ func main() {
 		base.Cfg.RoomServer.InternalAPI.Listen, // internal listener
 		setup.NoListener,                       // external listener
 		nil, nil,
+		base.Cfg.RoomServer.Database,
```
Comment (Kegsay) [inline cmd/dendrite-room-server/main.go:39]: These files no longer exist, please update to the lastest `master`.

**Review by Kegsay — CHANGES_REQUESTED** [review]
I would prefer to let each component decide when it is "up" than rely on pinging DBs - which may miss certain failure modes.

File: internal/httputil/healthcheck.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	"net/http"
+
+	"github.com/matrix-org/dendrite/internal/config"
+	"github.com/matrix-org/dendrite/internal/sqlutil"
+	"github.com/sirupsen/logrus"
+)
+
+// healthResponse is returned on requests to /api/health
+type healthResponse struct {
+	Code       int    `json:"code"`
```
[the remark this answers is not in the document]
Comment (S7evinK) [reply internal/httputil/healthcheck.go:30]: At least for Docker and Kubernetes the HTTP code is enough.
Outputting the actual error was more for other systems which might also check that endpoint.

File: internal/httputil/healthcheck.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	FirstError string `json:"error"`
+}
+
+// HealthCheckHandler adds a /health endpoint to the internal api mux
+func HealthCheckHandler(dbConfig ...config.DatabaseOptions) http.HandlerFunc {
+	conns := make([]*sql.DB, len(dbConfig))
+	// populate sql connections
+	for i, conf := range dbConfig {
+		c, err := sqlutil.Open(&conf)
```
[the remark this answers is not in the document]
Comment (S7evinK) [reply internal/httputil/healthcheck.go:39]: I'm not sure i understand.
Where do you see new connection pools being created? As far as i can see, the components are initialized first (db connections, consumers/producers, etc.) and only after that the healtcheck is added. (Which would reuse the already open connection; using one connection from the pool, when a `GET /api/health` is executed)
If anything failed during initialization, the component wouldn't reach the healtcheck part.

File: internal/sqlutil/trace.go (hunk 1 above)
Comment (S7evinK) [reply internal/sqlutil/trace.go]: Well, at least `roomserver/roomserver_test.go` reuses the same sqlite file for testing. So the approach with just using the `dbProperties.ConnectionString` as the key didn't work. (fixed now)

File: internal/httputil/healthcheck.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kegsay) [reply internal/httputil/healthcheck.go:39]: Sorry, I forgot that you modified `sqlutil.Open` to have a global shared map keyed off the DSN for opening `*sql.DB` instances.

**Review by Kegsay — CHANGES_REQUESTED** [review]
I like the idea of having a `/health` check, which is particularly important for polylith components which don't have `/_matrix/client/versions`.

I would prefer it if, for now, we just had a simple HTTP handler which responded with 200 OK in response to `/health`, without the DB pinging. I don't like the current pinging implementation because it adds state and side effects to `sqlutil.Open` making it much harder to reason about. I didn't like the previous implementation due to the creation of a new connection pool purely for `/health`. I'm unconvinced that we need to ping the DB in `/health` at present, given the predominant use case for `/health` is in k8s/ECS which the management system uses determine when the process is actually up. This does mean that if the DB were to disappear then the `/health` would not reflect this, but for now that's a tradeoff I'm happy with. We can revisit this if this actually becomes a problem.

Basically can we just have:
```go
b.InternalAPIMux.HandleFunc("/health", func(w http.ResponseWriter, req *http.Request) {
    w.WriteHeader(200)
})
```
in `internal/setup/base.go` and we're good to go.

**Comment by S7evinK** [comment]
You're probably right about pinging the database, as `database/sql` should reopen the connection when using `QueryContext`, so we shouldn't get into the situation where the server "has gone away".
I'll close this and try to come up with an implementation to at least check the availability of other required components (simple `net.DialTimeout`, which could also be used for e.g. Kafka)
